File: src/processing/technical_indicators.py
"""Módulo de cálculo de indicadores técnicos cuantitativos.

Proporciona cálculos vectorizados de alto rendimiento con pandas para:
- RSI (Relative Strength Index) con suavizado exponencial tipo Wilder.
- Bandas de Bollinger (media móvil central +/- n desviaciones estándar).
- MACD (Moving Average Convergence Divergence) y su histograma.
"""

from __future__ import annotations

import numpy as np
import pandas as pd


class TechnicalIndicators:
    """Calcula indicadores técnicos de momentum y volatilidad sobre series de precios."""

    @staticmethod
    def add_rsi(
        df: pd.DataFrame,
        price_col: str = "price_usd",
        period: int = 14,
    ) -> pd.DataFrame:
        """Calcula el Índice de Fuerza Relativa (RSI) y añade la columna 'rsi'.

        RSI = 100 - (100 / (1 + RS))
        Donde RS = Ganancia Promedio / Pérdida Promedio (usando suavizado Wilder / EWM).
        """
        df_out = df.copy()
        delta = df_out[price_col].diff()

        # Separar ganancias y pérdidas
        gain = delta.clip(lower=0)
        loss = -delta.clip(upper=0)

        # Suavizado exponencial estándar (Wilder: alpha = 1 / period)
        avg_gain = gain.ewm(alpha=1.0 / period, min_periods=period, adjust=False).mean()
        avg_loss = loss.ewm(alpha=1.0 / period, min_periods=period, adjust=False).mean()

        # Evitar división por cero
        rs = avg_gain / avg_loss.replace(0, np.nan)
        df_out["rsi"] = 100.0 - (100.0 / (1.0 + rs))

        # En caso de que avg_loss sea 0, RSI es 100
        df_out.loc[avg_loss == 0, "rsi"] = 100.0
        return df_out

    @staticmethod
    def add_bollinger_bands(
        df: pd.DataFrame,
        price_col: str = "price_usd",
        period: int = 20,
        num_std: float = 2.0,
    ) -> pd.DataFrame:
        """Calcula las Bandas de Bollinger y añade 'bb_middle', 'bb_upper', 'bb_lower'.

        - Banda Central = SMA(period)
        - Banda Superior = Central + (num_std * Desviación Estándar)
        - Banda Inferior = Central - (num_std * Desviación Estándar)
        """
        df_out = df.copy()
        middle = df_out[price_col].rolling(window=period).mean()
        std = df_out[price_col].rolling(window=period).std()

        df_out["bb_middle"] = middle
        df_out["bb_upper"] = middle + (num_std * std)
        df_out["bb_lower"] = middle - (num_std * std)
        df_out["bb_bandwidth"] = (df_out["bb_upper"] - df_out["bb_lower"]) / middle
        return df_out

    @staticmethod
    def add_macd(
        df: pd.DataFrame,
        price_col: str = "price_usd",
        fast_period: int = 12,
        slow_period: int = 26,
        signal_period: int = 9,
    ) -> pd.DataFrame:
        """Calcula el MACD y añade 'macd_line', 'macd_signal', 'macd_hist'.

        - Línea MACD = EMA(fast) - EMA(slow)
        - Línea Señal = EMA(signal) de Línea MACD
        - Histograma = Línea MACD - Línea Señal
        """
        df_out = df.copy()
        ema_fast = df_out[price_col].ewm(span=fast_period, adjust=False).mean()
        ema_slow = df_out[price_col].ewm(span=slow_period, adjust=False).mean()

        macd_line = ema_fast - ema_slow
        macd_signal = macd_line.ewm(span=signal_period, adjust=False).mean()
        macd_hist = macd_line - macd_signal

        df_out["macd_line"] = macd_line
        df_out["macd_signal"] = macd_signal
        df_out["macd_hist"] = macd_hist
        return df_out

    @classmethod
    def add_all_indicators(cls, df: pd.DataFrame) -> pd.DataFrame:
        """Aplica todos los indicadores técnicos estándar al DataFrame."""
        df_out = cls.add_rsi(df)
        df_out = cls.add_bollinger_bands(df_out)
        df_out = cls.add_macd(df_out)
        return df_out

File: src/lib/technical_indicators.ts
import { ProcessedMarketPoint } from '../types';

export class TechnicalIndicators {
  /**
   * Exponential Moving Average calculation.
   */
  static calcEMA(values: number[], span: number): number[] {
    const alpha = 2 / (span + 1);
    const ema: number[] = new Array(values.length);
    if (values.length === 0) return ema;

    ema[0] = values[0];
    for (let i = 1; i < values.length; i++) {
      ema[i] = values[i] * alpha + ema[i - 1] * (1 - alpha);
    }
    return ema;
  }

  /**
   * Calculates the Relative Strength Index (RSI) using Wilder's smoothing.
   */
  static addRSI(data: ProcessedMarketPoint[], period = 14): ProcessedMarketPoint[] {
    const n = data.length;
    if (n === 0) return [];

    const result = data.map((d) => ({ ...d }));
    const alpha = 1.0 / period;

    let avgGain = 0;
    let avgLoss = 0;

    // Initial period
    for (let i = 1; i < Math.min(period, n); i++) {
      const diff = result[i].price_usd - result[i - 1].price_usd;
      if (diff >= 0) avgGain += diff;
      else avgLoss += -diff;
    }
    avgGain /= period;
    avgLoss /= period;

    for (let i = 0; i < n; i++) {
      if (i < period) {
        result[i].rsi = undefined;
        continue;
      }

      const diff = result[i].price_usd - result[i - 1].price_usd;
      const gain = diff > 0 ? diff : 0;
      const loss = diff < 0 ? -diff : 0;

      avgGain = avgGain * (1 - alpha) + gain * alpha;
      avgLoss = avgLoss * (1 - alpha) + loss * alpha;

      if (avgLoss === 0) {
        result[i].rsi = 100.0;
      } else {
        const rs = avgGain / avgLoss;
        result[i].rsi = 100.0 - 100.0 / (1.0 + rs);
      }
    }

    return result;
  }

  /**
   * Calculates Bollinger Bands (middle, upper, lower, bandwidth).
   */
  static addBollingerBands(
    data: ProcessedMarketPoint[],
    period = 20,
    numStd = 2.0
  ): ProcessedMarketPoint[] {
    const result = data.map((d) => ({ ...d }));
    const n = result.length;

    for (let i = 0; i < n; i++) {
      if (i < period - 1) {
        result[i].bb_middle = undefined;
        result[i].bb_upper = undefined;
        result[i].bb_lower = undefined;
        result[i].bb_bandwidth = undefined;
        continue;
      }

      let sum = 0;
      for (let j = i - period + 1; j <= i; j++) {
        sum += result[j].price_usd;
      }
      const mean = sum / period;

      let varSum = 0;
      for (let j = i - period + 1; j <= i; j++) {
        varSum += Math.pow(result[j].price_usd - mean, 2);
      }
      const std = Math.sqrt(varSum / (period - 1 || 1));

      result[i].bb_middle = mean;
      result[i].bb_upper = mean + numStd * std;
      result[i].bb_lower = mean - numStd * std;
      result[i].bb_bandwidth = (result[i].bb_upper! - result[i].bb_lower!) / (mean || 1);
    }

    return result;
  }

  /**
   * Calculates MACD (line, signal, histogram).
   */
  static addMACD(
    data: ProcessedMarketPoint[],
    fastPeriod = 12,
    slowPeriod = 26,
    signalPeriod = 9
  ): ProcessedMarketPoint[] {
    const result = data.map((d) => ({ ...d }));
    const prices = result.map((d) => d.price_usd);

    const emaFast = this.calcEMA(prices, fastPeriod);
    const emaSlow = this.calcEMA(prices, slowPeriod);

    const macdLine: number[] = new Array(prices.length);
    for (let i = 0; i < prices.length; i++) {
      macdLine[i] = emaFast[i] - emaSlow[i];
    }

    const macdSignal = this.calcEMA(macdLine, signalPeriod);

    for (let i = 0; i < result.length; i++) {
      result[i].macd_line = macdLine[i];
      result[i].macd_signal = macdSignal[i];
      result[i].macd_hist = macdLine[i] - macdSignal[i];
    }

    return result;
  }

  /**
   * Applies all indicators to the series.
   */
  static addAllIndicators(data: ProcessedMarketPoint[]): ProcessedMarketPoint[] {
    let res = this.addRSI(data, 14);
    res = this.addBollingerBands(res, 20, 2.0);
    res = this.addMACD(res, 12, 26, 9);
    return res;
  }
}
